modules/security-jwt/src/main/java/org/opencastproject/security/jwt/GuavaCachedUrlJwkProvider.java

```diff
@@ -0,0 +1,148 @@
+/**
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ *
+ * The Apereo Foundation licenses this file to you under the Educational
+ * Community License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License
+ * at:
+ *
+ *   http://opensource.org/licenses/ecl2.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.opencastproject.security.jwt;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+import com.auth0.jwk.Jwk;
+import com.auth0.jwk.JwkException;
+import com.auth0.jwk.SigningKeyNotFoundException;
+import com.auth0.jwk.UrlJwkProvider;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * JWK provider that caches previously fetched JWKs in memory using a Google Guava cache.
+ */
+public class GuavaCachedUrlJwkProvider extends UrlJwkProvider {
+
+  /** Logging facility. */
+  private static final Logger logger = LoggerFactory.getLogger(GuavaCachedUrlJwkProvider.class);
+
+  /** JWK cache. */
+  private final Cache<String, List<Jwk>> cache;
+
+  /** Key for the JWK in the cache. */
+  private static final String KEY = "GET_ALL";
+
+  /**
+   * Cons a new cached provider from a JWKs URL and a TTL.
+   *
+   * @param jwksUrl The URL where JWKs are published.
+   * @param expiresIn The amount of time the fetched JWKs will live in the cache.
+   * @param expiresUnit The unit of the expiresIn parameter.
+   */
+  public GuavaCachedUrlJwkProvider(String jwksUrl, long expiresIn, TimeUnit expiresUnit) {
+    super(urlFromString(jwksUrl));
+    this.cache = CacheBuilder.newBuilder().maximumSize(1).expireAfterWrite(expiresIn, expiresUnit).build();
```

Comment:
What's the logic here behind this having a cache of size 1, when the above code only supports a single url?

Reply:
I needed an expiring cache since this URL should be re-fetched regularly. Guava does not offer an option for single-element caches (see google/guava#872). This was the easiest way to implement this as far as I know.